manager/controlapi/volume.go

```diff
@@ -0,0 +1,226 @@
+package controlapi
+
+import (
+	"context"
+
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/identity"
+	"github.com/docker/swarmkit/manager/state/store"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func (s *Server) CreateVolume(ctx context.Context, request *api.CreateVolumeRequest) (*api.CreateVolumeResponse, error) {
+	if request.Spec == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "spec must not be nil")
+	}
+
+	// validate the volume spec
+	if request.Spec.Driver == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "driver must be specified")
+	}
+
+	if request.Spec.Annotations.Name == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "meta: name must be provided")
+	}
+
+	volume := &api.Volume{
+		ID:   identity.NewID(),
+		Spec: *request.Spec,
+	}
+	err := s.store.Update(func(tx store.Tx) error {
+		// verify that the specified driver exists
+		clusters, err := store.FindClusters(tx, store.ByName(store.DefaultClusterName))
+		if err == nil && len(clusters) == 1 {
+			found := false
+			for _, plugin := range clusters[0].Spec.CSIConfig.Plugins {
+				if plugin.Name == volume.Spec.Driver.Name {
+					found = true
+					break
+				}
+			}
+
+			if !found {
+				return status.Errorf(
+					codes.FailedPrecondition,
+					"CSI plugin %v not found", volume.Spec.Driver.Name,
+				)
+			}
+		}
+
+		for _, secret := range volume.Spec.Secrets {
+			s := store.GetSecret(tx, secret.Secret)
+			if s == nil {
+				return status.Errorf(codes.FailedPrecondition, "secret %v does not exist", secret.Secret)
```

Comment:
Since you can have multiple secrets, does it make sense to provide an error to the caller of all the secrets that don't exist, instead of just the first?

Reply:
Actually, yes. I'll go back and change this.